pkg/api/norman/store/cluster/cluster_store.go

```diff
@@ -544,21 +544,34 @@ func (r *Store) Update(apiContext *types.APIContext, schema *types.Schema, data
 		return nil, err
 	}
 
-	// Replace rancherKubernetesEngineConfig.cloudProvider.vsphereCloudProvider.virtualCenter value to properly update virtualCenter removal
+	// Replace rancherKubernetesEngineConfig.cloudProvider.vsphereCloudProvider.virtualCenter value to properly update if any virtualcenter has been removed
 	if newVCenter, ok := values.GetValue(data, "rancherKubernetesEngineConfig", "cloudProvider", "vsphereCloudProvider", "virtualCenter"); ok && newVCenter != nil {
-		oldVCenter, ok := values.GetValue(existingCluster, "rancherKubernetesEngineConfig", "cloudProvider", "vsphereCloudProvider", "virtualCenter")
-		if ok && oldVCenter != nil && !reflect.DeepEqual(newVCenter, oldVCenter) {
-			newData := map[string]interface{}{}
-			for k, v := range data {
-				newData[k] = v
-			}
-			// Update virtualCenter value to nil
-			values.PutValue(newData, nil, "rancherKubernetesEngineConfig", "cloudProvider", "vsphereCloudProvider", "virtualCenter")
-			if _, err := r.Store.Update(apiContext, schema, newData, id); err != nil {
-				return nil, err
+		if oldVCenter, ok := values.GetValue(existingCluster, "rancherKubernetesEngineConfig", "cloudProvider", "vsphereCloudProvider", "virtualCenter"); ok && oldVCenter != nil {
+			if oldVCenterMap, oldOk := oldVCenter.(map[string]interface{}); oldOk && oldVCenterMap != nil {
+				if newVCenterMap, newOk := newVCenter.(map[string]interface{}); newOk && newVCenterMap != nil {
+					oldRowRemoved := false
+					for k := range oldVCenterMap {
+						if _, ok := newVCenterMap[k]; !ok && oldVCenterMap[k] != nil {
+							oldRowRemoved = true
+							break
+						}
+					}
+					// Update just if any virtualcenter has been removed
+					if oldRowRemoved {
+						newData := map[string]interface{}{}
+						for k, v := range data {
+							newData[k] = v
+						}
+						// Update virtualCenter value to nil
+						values.PutValue(newData, nil, "rancherKubernetesEngineConfig", "cloudProvider", "vsphereCloudProvider", "virtualCenter")
+						if _, err := r.Store.Update(apiContext, schema, newData, id); err != nil {
```

Comment:
@superseb , you are right, this is causing the cluster to go into provision state twice. But the root cause of the issue seems to be at the agent restart more than the second update. Anyway, here i'm thinking to add a cluster annotation to avoid the first cluster update, removing it when come bak and update again, what do you think??

Reply:
@superseb , I've added a `io.cattle.rke.no.update` cluster annotation to skip for `r.Store.Update`. Could you please take a view??

Reply:
Just to be clear here, this approach is needed because with the current API behavior, we cannot delete an entry because it will merge the given input, correct? So to work around this, we empty the value (first `Store.Update`), and then put the correct value (second/last `Store.Update`). Because `Store.Update` will trigger cluster provisioning, we add an annotation to make sure that it doesn't trigger provisioning on the first `Store.Update`. And this approach was preferred because we didn't want to change the API because of scope of impact/unexpected behavioral change?

Reply:
That's it.